CRM/Utils/File.php

```diff
@@ -766,15 +771,36 @@ public static function findFiles($dir, $pattern, $relative = FALSE) {
         }
       }
       // Find subdirs to recurse into.
-      $subdirs = glob("$subdir/*", GLOB_ONLYDIR);
-      if (!empty($excludeDirsPattern)) {
-        $subdirs = preg_grep($excludeDirsPattern, $subdirs, PREG_GREP_INVERT);
+      $depth = static::findPathDepth($subdir) - $baseDepth + 1;
+      if ($maxDepth === NULL || $depth <= $maxDepth) {
+        $subdirs = glob("$subdir/*", GLOB_ONLYDIR);
+        if (!empty($excludeDirsPattern)) {
+          $subdirs = preg_grep($excludeDirsPattern, $subdirs, PREG_GREP_INVERT);
+        }
+        $todos = array_merge($todos, $subdirs);
       }
-      $todos = array_merge($todos, $subdirs);
     }
     return $result;
   }
 
+  /**
+   * Determine the absolute depth of a path expression.
+   *
+   * @param string $path
+   *   Ex: '/var/www/foo'
+   * @return int
+   *   Ex: 3
+   */
+  private static function findPathDepth(string $path): int {
+    // Both PHP-Unix and PHP-Windows support '/'s. Additionally, PHP-Windows also supports '\'s.
+    // They are roughly equivalent. (The differences are described by a secret book hidden in the tower of Mordor.)
+    $depth = substr_count($path, '/');
+    if (DIRECTORY_SEPARATOR !== '/') {
+      $depth += substr_count($path, PATH_SEPARATOR);
```

Comment:
This should be DIRECTORY_SEPARATOR instead of PATH.

But the bigger question is maybe who is going to use this setting, and, if it should exist, then could an upgrade script calculate what it should be based on the state of the system at the time (extensions get installed and uninstalled, but the depth is unlikely to increase except for a very weird extension).

Also I get an error if I set the setting to 2: `Uncaught Error: Class 'Civi\Api4\Subscriber\AutocompleteSubscriber' not found in ...\ext\afform\core\afform.php:59`. So it suggests it shouldn't be allowed to be less than 3.

As noted it is a bit difficult to evaluate the effect. I'm not sure where efforts should focus. We have a couple suggestions scattered a bit:
* Serialize is slow
* The change in the class scanning PR.
* excessive cache clearing
* Does it need to scan anything (maybe it's convenient, but has a trade-off)